pkg/client/outofbandv2/client.go

```diff
@@ -0,0 +1,200 @@
+type Event interface {
+	// Error is non-nil if an error is encountered.
+	Error() error
+}
+
+// MessageOption allow you to customize the way out-of-band messages are built.
+type MessageOption func(*message)
+
+type message struct {
+	Label              string
+	Goal               string
+	GoalCode           string
+	RouterConnections  []string
+	Service            []interface{}
+	HandshakeProtocols []string
+	Attachments        []*decorator.AttachmentV2
+	Accept             []string
+}
+
+// OobService defines the outofband service.
+type OobService interface {
+	service.Event
+	AcceptInvitation(*oobv2.Invitation, oobv2.Options) error
+	Actions() ([]oobv2.Action, error)
+	ActionContinue(string, oobv2.Options) error
+	ActionStop(string, error) error
+}
+
+// Provider provides the dependencies for the client.
+type Provider interface {
+	ServiceEndpoint() string
+	Service(id string) (interface{}, error)
+	KMS() kms.KeyManager
+	KeyType() kms.KeyType
+	KeyAgreementType() kms.KeyType
+	MediaTypeProfiles() []string
+}
+
+// Client for the Out-Of-Band protocol:
+// https://github.com/hyperledger/aries-rfcs/blob/master/features/0434-outofband/README.md
+type Client struct {
+	service.Event
+	oobService        OobService
+	mediaTypeProfiles []string
+}
+
+// New returns a new Client for the Out-Of-Band protocol.
+func New(p Provider) (*Client, error) {
+	s, err := p.Service(oobv2.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to look up service %s : %w", oobv2.Name, err)
+	}
+
+	oobSvc, ok := s.(OobService)
+	if !ok {
+		return nil, fmt.Errorf("failed to cast service %s as a dependency", oobv2.Name)
+	}
+
+	mtp := p.MediaTypeProfiles()
+
+	if len(mtp) == 0 {
+		mtp = []string{transport.MediaTypeDIDCommV2Profile}
+	}
+
+	client := &Client{
+		Event:             oobSvc,
+		oobService:        oobSvc,
+		mediaTypeProfiles: mtp,
+	}
+
+	return client, nil
+}
+
+// CreateInvitation creates and saves an out-of-band/v2 invitation.
+func (c *Client) CreateInvitation(opts ...MessageOption) (*oobv2.Invitation, error) {
+	msg := &message{}
+
+	for _, opt := range opts {
+		opt(msg)
+	}
+
+	inv := &oobv2.Invitation{
+		ID:    uuid.New().String(),
+		Type:  InvitationMsgType,
+		Label: msg.Label,
+		Body: &oobv2.InvitationBody{
+			Goal:     msg.Goal,
+			GoalCode: msg.GoalCode,
+			Accept:   msg.Accept,
+		},
+		Requests: msg.Attachments,
+	}
+
+	if len(inv.Body.Accept) == 0 {
+		inv.Body.Accept = c.mediaTypeProfiles
+	}
+
+	return inv, nil
+}
+
+// Actions returns unfinished actions for the async usage.
+func (c *Client) Actions() ([]oobv2.Action, error) {
+	actions, err := c.oobService.Actions()
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]oobv2.Action, len(actions))
+	copy(result, actions)
+
+	return result, nil
+}
+
+// ActionContinue allows continuing with the protocol after an action event was triggered.
+func (c *Client) ActionContinue(piID string, opts oobv2.Options) error {
```

Comment:
When does a client need to invoke this function ? Do they have to call `AcceptInvitation` first and then invoke this ?

Reply:
so ActionContinue is supposed to trigger the other services, would it be the case? If it's not needed, i can definitely remove it. We just need to figure out where/how to trigger the other protocol services via oob.

Reply:
removed ActionContinue(), Actions() and ActionStop()